siatest/renterhost/renterhost_test.go

```diff
@@ -0,0 +1,100 @@
+package renterhost
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/NebulousLabs/Sia/siatest"
+)
+
+// TestRenterHost executes a number of subtests using the same TestGroup to
+// save time on initialization
+func TestRenterHost(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+
+	// Create a group for the subtests
+	groupParams := siatest.GroupParams{
+		Hosts:   5,
+		Renters: 1,
+		Miners:  1,
+	}
+	tg, err := siatest.NewGroupFromTemplate(groupParams)
+	if err != nil {
+		t.Fatal("Failed to create group: ", err)
+	}
+	defer func() {
+		if err := tg.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}()
+	// Specifiy subtests to run
+	subTests := []struct {
+		name string
+		test func(*testing.T, *siatest.TestGroup)
+	}{
+		{"UploadDownload", testUploadDownload},
+	}
+	// Run subtests
+	for _, subtest := range subTests {
+		t.Run(subtest.name, func(t *testing.T) {
+			subtest.test(t, tg)
+		})
+	}
+}
+
+// testUploadDownload is a subtest that uses an existing TestGroup to test if
+// uploading and downloading a file works
+func testUploadDownload(t *testing.T, tg *siatest.TestGroup) {
+	// Grab the first of the group's renters
+	renter := tg.Renters()[0]
+	// Create file for upload
+	file, err := siatest.NewFile(100)
+	if err != nil {
+		t.Fatal("Failed to create file for testing: ", err)
+	}
+	// Upload file, creating a parity piece for each host in the group
+	err = renter.Upload(file, 1, uint64(len(tg.Hosts())))
+	if err != nil {
+		t.Fatal("Failed to start upload: ", err)
+	}
+	// Wait until upload reached 100% progress
+	if err := renter.WaitForUploadProgress(file, 1); err != nil {
+		t.Error(err)
+	}
+	// Wait until upload reaches len(tg.Hosts()) redundancy
+	if err := renter.WaitForUploadRedundancy(file, float64(len(tg.Hosts()))); err != nil {
+		t.Error(err)
+	}
+	// Download the file synchronously directly into memory and compare the
+	// data to the original
+	downloadedData, err := renter.Download(file)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if bytes.Compare(downloadedData, file.Bytes()) != 0 {
```

Comment:
The compare piece can probably be a part of the `Download` call, makes sense to me that you would automatically check the integrity of the file after you download it every single time in a testing environment. Doesn't even need to be a separate function, as I can't imagine a situation where checking the integrity of a file after downloading it would be unwanted (unlike waiting for an upload to hit full redundancy, where it's easy to see testing situations where you don't want the file to hit full redundancy).

Reply:
I am not sure it can. `TestFile.Download(...)` doesn't require the file to be stored locally. In the future we might add calls like `TestFile.Delete(...)` to delete the local copy of a file. This means, that we might not have data to compare the download to.
We could store a hash in the `TestFile` object for that purpose but that still would allow us to check the contents when we use blocking download calls.